src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs

```diff
@@ -216,26 +217,137 @@ private static unsafe string[] GetCommandLineArgsNative()
             char* lpCmdLine = Interop.Kernel32.GetCommandLine();
             Debug.Assert(lpCmdLine != null);
 
-            int numArgs = 0;
-            char** argvW = Interop.Shell32.CommandLineToArgv(lpCmdLine, &numArgs);
-            if (argvW == null)
-            {
-                ThrowHelper.ThrowOutOfMemoryException();
-            }
+            return SegmentCommandLine(lpCmdLine);
+        }
 
-            try
+        private static unsafe string[] SegmentCommandLine(char* cmdLine)
+        {
+            //---------------------------------------------------------------------
+            // Splits a command line into argc/argv lists, using the VC7 parsing rules.
+            //
+            // This functions interface mimics the CommandLineToArgvW api.
+            //
+            //---------------------------------------------------------------------
+            // NOTE: Implementation-wise, once every few years it would be a good idea to
+            // compare this code with the C runtime library's parse_cmdline method,
+            // which is in vctools\crt\crtw32\startup\stdargv.c.  (Note we don't
+            // support wild cards, and we use Unicode characters exclusively.)
+            // We are up to date as of ~6/2005.
+            //---------------------------------------------------------------------
+
+            // The C# version is ported from C++ version in coreclr
+            // The behavior mimics what MSVCRT does before main,
+            // which is slightly different with CommandLineToArgvW
+
+            ArrayBuilder<string> arrayBuilder = default;
```

Comment:
For reference, you can get to the current version of the universal CRT command line parser by creating a simple C++ command line application in Visual Studio, setting a break point at `ucrtbased!parse_command_line<char>` and then selecting "Go To Source Code" to make Visual Studio download the source code. It still looks the same.

Reply:
The current version is identical for UTF-16. It seems that the source link is Microsoft-only right? Don't confuse external contributors who read this.

Reply:
UCRT source link should be public. These steps worked for me even when I was not authenticated using my Microsoft account. Also, https://learn.microsoft.com/en-us/cpp/c-runtime-library/crt-debugging-techniques suggests that the UCRT sources are publically available for debugging purposes.